docs/tutorials/rerank/rerank_pipeline_with_bge-rerank-m3-v2_model_deployed_on_Sagemaker.md

```diff
@@ -152,9 +189,46 @@ POST /_plugins/_ml/connectors/_create
       "headers": {
         "content-type": "application/json"
       },
+      "pre_process_function": """
+        def query_text = params.query_text;
+        def text_docs = params.text_docs;
+        def textDocsBuilder = new StringBuilder('[');
+        for (int i=0; i<text_docs.length; i++) {
+          textDocsBuilder.append('\"');
+          textDocsBuilder.append(text_docs[i]);
+          textDocsBuilder.append('\"');
+          if (i<text_docs.length - 1) {
+            textDocsBuilder.append(',');
+          }
+        }
+        textDocsBuilder.append(']');
+        def parameters = '{ \"query\": \"' + query_text + '\",  \"texts\": ' + textDocsBuilder.toString() + ' }';
+        return  '{\"parameters\": ' + parameters + '}';
+      """,
       "request_body": "{ \"query\": \"${parameters.query}\", \"texts\": ${parameters.texts} }",
-      "pre_process_function": "\n    def query_text = params.query_text;\n    def text_docs = params.text_docs;\n    def textDocsBuilder = new StringBuilder('[');\n    for (int i=0; i<text_docs.length; i++) {\n      textDocsBuilder.append('\"');\n      textDocsBuilder.append(text_docs[i]);\n      textDocsBuilder.append('\"');\n      if (i<text_docs.length - 1) {\n        textDocsBuilder.append(',');\n      }\n    }\n    textDocsBuilder.append(']');\n    def parameters = '{ \"query\": \"' + query_text + '\",  \"texts\": ' + textDocsBuilder.toString() + ' }';\n    return  '{\"parameters\": ' + parameters + '}';\n",
-      "post_process_function": "\n      \n      def dataType = \"FLOAT32\";\n      \n      \n      if (params.result == null)\n      {\n          return 'no result generated';\n          //return params.response;\n      }\n      def outputs = params.result;\n      \n      def sorted_outputs = outputs;\n      for (int i=0; i<outputs.length; i++) {\n        def idx = new BigDecimal(outputs[i].index.toString()).intValue();\n        sorted_outputs[idx] = outputs[i];\n      }\n      def resultBuilder = new StringBuilder('[');\n      for (int i=0; i<outputs.length; i++) {\n        resultBuilder.append(' {\"name\": \"similarity\", \"data_type\": \"FLOAT32\", \"shape\": [1],');\n        //resultBuilder.append('{\"name\": \"similarity\"}');\n        \n        resultBuilder.append('\"data\": [');\n        resultBuilder.append(outputs[i].score);\n        resultBuilder.append(']}');\n        if (i<outputs.length - 1) {\n          resultBuilder.append(',');\n        }\n      }\n      resultBuilder.append(']');\n      \n      return resultBuilder.toString();\n    "
+      "post_process_function": """
+        if (params.result == null) {
+          return "no result generated";

```

Comment:
should here throw exception rather than returning the string?

Reply:
You are correct, throwing exception is prefer than just returning string. As far as I can see, pre-built function throws exceptions for these situations.

- https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/connector/functions/postprocess/CohereRerankPostProcessFunction.java

I'd like to know implementation rule for pre or post processing functions. Do you know coding style or rule on the function? Other sample looks not include codes to through exceptions.

- https://opensearch.org/docs/2.17/ml-commons-plugin/remote-models/connectors/#amazon-bedrock-connector

Reply:
you should write the exception in painless. 

maybe try this way?

```
if (params.result == null) {
    throw new IllegalArgumentException("Result is null");
}
```

Reply:
I've tested with throwing exception. When embedding throwing exception, client sees following results. It's vague output but better than returning 200 code.

```
TransportError: TransportError(500, 'm_l_exception', 'runtime error')
```

```
{
  "inference_results": [
    {
      "output": [
        {
          "name": "response",
          "dataAsMap": {
            "response": "no result generated"
          }
        }
      ],
      "status_code": 200
    }
  ]
}
```

I'll update to replace error handling.